File: backend/src/services/skill_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from src.model.skill import Skill, SkillCreate, SkillResponse
from typing import List, Optional
import logging

logger = logging.getLogger(__name__)

class SkillService:
    """Service class for handling skill-related business logic."""
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_skill(self, skill_data: SkillCreate) -> SkillResponse:
        try:
            db_skill = Skill(**skill_data.model_dump())
            self.db.add(db_skill)
            self.db.commit()
            self.db.refresh(db_skill)
            logger.info(f"Created skill with ID: {db_skill.id}")
            return SkillResponse.model_validate(db_skill)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to create skill due to integrity error: {e}")
            raise ValueError("Skill creation failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to create skill: {e}")
            raise ValueError(f"Skill creation failed: {str(e)}")
    
    def get_skill_by_id(self, skill_id: int) -> Optional[SkillResponse]:
        skill = self.db.query(Skill).filter(Skill.id == skill_id).first()
        if skill:
            return SkillResponse.model_validate(skill)
        return None
    
    def list_skills(self, skip: int = 0, limit: int = 10) -> List[SkillResponse]:
        skills = self.db.query(Skill).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(skills)} skills with skip={skip}, limit={limit}")
        return [SkillResponse.model_validate(skill) for skill in skills]
    
    def update_skill(self, skill_id: int, skill_data: SkillCreate) -> Optional[SkillResponse]:
        db_skill = self.db.query(Skill).filter(Skill.id == skill_id).first()
        if not db_skill:
            return None
        try:
            for key, value in skill_data.model_dump().items():
                setattr(db_skill, key, value)
            self.db.commit()
            self.db.refresh(db_skill)
            logger.info(f"Updated skill with ID: {db_skill.id}")
            return SkillResponse.model_validate(db_skill)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to update skill {skill_id} due to integrity error: {e}")
            raise ValueError("Update failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to update skill {skill_id}: {e}")
            raise ValueError(f"Update failed: {str(e)}")
    
    def delete_skill(self, skill_id: int) -> bool:
        db_skill = self.db.query(Skill).filter(Skill.id == skill_id).first()
        if not db_skill:
            return False
        try:
            # Delete related user_skills relationships
            from src.model.user_skill import UserSkill
            user_skills = self.db.query(UserSkill).filter(UserSkill.skill_id == skill_id).all()
            for user_skill in user_skills:
                self.db.delete(user_skill)
            
            # Delete related project_skills relationships
            from src.model.project_skill import ProjectSkill
            project_skills = self.db.query(ProjectSkill).filter(ProjectSkill.skill_id == skill_id).all()
            for project_skill in project_skills:
                self.db.delete(project_skill)
            
            # Delete the skill itself
            self.db.delete(db_skill)
            self.db.commit()
            logger.info(f"Deleted skill with ID {skill_id} and {len(user_skills)} user relationships and {len(project_skills)} project relationships")
            return True
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to delete skill {skill_id}: {e}")
            raise ValueError(f"Deletion failed: {str(e)}")
    
    def get_skills_count(self) -> int:
        return self.db.query(Skill).count()
    
    def search_skills(self, search_term: str, skip: int = 0, limit: int = 10) -> List[SkillResponse]:
        search_pattern = f"%{search_term}%"
        skills = self.db.query(Skill).filter(
            Skill.name.ilike(search_pattern)
        ).offset(skip).limit(limit).all()
        logger.info(f"Found {len(skills)} skills matching search term: {search_term}")
        return [SkillResponse.model_validate(skill) for skill in skills]
    
    def bulk_create_skills(self, skills_data: List[SkillCreate]) -> List[SkillResponse]:
        """Create multiple skills in a single transaction."""
        created_skills = []
        try:
            for skill_data in skills_data:
                db_skill = Skill(**skill_data.model_dump())
                self.db.add(db_skill)
                created_skills.append(db_skill)
            
            self.db.commit()
            
            # Refresh all created skills
            for skill in created_skills:
                self.db.refresh(skill)
            
            logger.info(f"Bulk created {len(created_skills)} skills")
            return [SkillResponse.model_validate(skill) for skill in created_skills]
            
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to bulk create skills due to integrity error: {e}")
            raise ValueError("Bulk skill creation failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to bulk create skills: {e}")
            raise ValueError(f"Bulk skill creation failed: {str(e)}") 

File: frontend/src/services/skillService.js
import apiClient, {
  API_SKILLS,
  API_SKILLS_COUNT,
  API_SKILLS_SEARCH,
  API_SKILLS_BULK,
  API_SKILL_BY_ID,
} from './api';

class SkillService {
  // Create a new skill
  async createSkill(skillData) {
    try {
      console.log('SkillService: Creating skill with data:', skillData);
      const response = await apiClient.post(API_SKILLS, skillData);
      console.log('SkillService: Create response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error creating skill:', error);
      console.error('Error response:', error.response?.data);
      throw error;
    }
  }

  // Create multiple skills at once
  async createMultipleSkills(skillsData) {
    try {
      console.log('SkillService: Creating multiple skills with data:', skillsData);
      const response = await apiClient.post(API_SKILLS_BULK, skillsData);
      console.log('SkillService: Bulk create response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error creating multiple skills:', error);
      console.error('Error response:', error.response?.data);
      throw error;
    }
  }

  // Get all skills with pagination
  async getSkills(skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_SKILLS, {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching skills:', error);
      throw error;
    }
  }

  // Get skill by ID
  async getSkillById(skillId) {
    try {
      const response = await apiClient.get(API_SKILL_BY_ID(skillId));
      return response.data;
    } catch (error) {
      console.error('Error fetching skill:', error);
      throw error;
    }
  }

  // Update skill
  async updateSkill(skillId, skillData) {
    try {
      console.log('SkillService: Updating skill', skillId, 'with data:', skillData);
      const response = await apiClient.put(API_SKILL_BY_ID(skillId), skillData);
      console.log('SkillService: Update response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error updating skill:', error);
      console.error('Error response:', error.response?.data);
      throw error;
    }
  }

  // Delete skill
  async deleteSkill(skillId) {
    try {
      const response = await apiClient.delete(API_SKILL_BY_ID(skillId));
      return response.data;
    } catch (error) {
      console.error('Error deleting skill:', error);
      throw error;
    }
  }

  // Get skills count
  async getSkillsCount() {
    try {
      const response = await apiClient.get(API_SKILLS_COUNT);
      return response.data;
    } catch (error) {
      console.error('Error fetching skills count:', error);
      throw error;
    }
  }

  // Search skills
  async searchSkills(searchTerm, skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_SKILLS_SEARCH, {
        params: { q: searchTerm, skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error searching skills:', error);
      throw error;
    }
  }

  // Get skills for multi-select (formatted for dropdowns)
  async getSkillsForMultiSelect() {
    try {
      const response = await apiClient.get(`${API_SKILLS}/multi-select`);
      return response.data;
    } catch (error) {
      console.error('Error fetching skills for multi-select:', error);
      throw error;
    }
  }
}

export default new SkillService(); 